plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/AsyncUploadUtils.java

```diff
@@ -0,0 +1,366 @@
+        UploadRequest uploadRequest,
+        StreamContext streamContext
+    ) {
+
+        CompletableFuture<UploadResponse> returnFuture = new CompletableFuture<>();
+        try {
+            if (streamContext.getNumberOfParts() == 1) {
+                log.debug(() -> "Starting the upload as a single upload part request");
+                uploadInOneChunk(s3AsyncClient, uploadRequest, streamContext.getStreamProvider().provideStream(0), returnFuture);
+            } else {
+                log.debug(() -> "Starting the upload as multipart upload request");
+                uploadInParts(s3AsyncClient, uploadRequest, streamContext, returnFuture);
+            }
+        } catch (Throwable throwable) {
+            returnFuture.completeExceptionally(throwable);
+        }
+
+        return returnFuture;
+    }
+
+    private void uploadInParts(
+        S3AsyncClient s3AsyncClient,
+        UploadRequest uploadRequest,
+        StreamContext streamContext,
+        CompletableFuture<UploadResponse> returnFuture
+    ) {
+
+        CreateMultipartUploadRequest request = CreateMultipartUploadRequest.builder()
+            .bucket(uploadRequest.getBucket())
+            .key(uploadRequest.getKey())
+            .build();
+        CompletableFuture<CreateMultipartUploadResponse> createMultipartUploadFuture = SocketAccess.doPrivileged(
+            () -> s3AsyncClient.createMultipartUpload(request)
+        );
+
+        // Ensure cancellations are forwarded to the createMultipartUploadFuture future
+        CompletableFutureUtils.forwardExceptionTo(returnFuture, createMultipartUploadFuture);
+
+        createMultipartUploadFuture.whenComplete((createMultipartUploadResponse, throwable) -> {
+            if (throwable != null) {
+                handleException(returnFuture, () -> "Failed to initiate multipart upload", throwable);
+            } else {
+                log.debug(() -> "Initiated new multipart upload, uploadId: " + createMultipartUploadResponse.uploadId());
+                doUploadInParts(s3AsyncClient, uploadRequest, streamContext, returnFuture, createMultipartUploadResponse.uploadId());
+            }
+        });
+    }
+
+    private void doUploadInParts(
+        S3AsyncClient s3AsyncClient,
+        UploadRequest uploadRequest,
+        StreamContext streamContext,
+        CompletableFuture<UploadResponse> returnFuture,
+        String uploadId
+    ) {
+
+        // The list of completed parts must be sorted
+        AtomicReferenceArray<CompletedPart> completedParts = new AtomicReferenceArray<>(streamContext.getNumberOfParts());
+
+        List<CompletableFuture<CompletedPart>> futures;
+        try {
+            futures = sendUploadPartRequests(s3AsyncClient, uploadRequest, streamContext, uploadId, completedParts);
+        } catch (Exception ex) {
+            try {
+                cleanUpParts(s3AsyncClient, uploadRequest, uploadId);
+            } finally {
+                returnFuture.completeExceptionally(ex);
```

Comment:
we should do this before attempting `cleanUpParts`

Reply:
@gbbafna would the behaviour be different if we do that? Since it is in the finally block, it will always be completed exceptionally so long as `catch (Exception ex)` is invoked. What do you think?

Reply:
This is a failure scenario and I think its fine to handle clean up in sync.